crates/jiter/src/value.rs

```diff
@@ -257,157 +291,200 @@ fn take_value_recursive<'j, 's>(
-                            let n = parser
-                                .consume_number::<NumberAny>(peek.into_inner(), allow_inf_nan)
-                                .map_err(|e| {
-                                    if !peek.is_num() {
-                                        json_error!(ExpectedSomeValue, parser.index)
-                                    } else {
-                                        e
-                                    }
-                                })?;
-                            match n {
+                        _ => parser
+                            .consume_number::<NumberAny>(peek.into_inner(), allow_inf_nan)
+                            .map_err(|e| {
+                                if !peek.is_num() {
+                                    json_error!(ExpectedSomeValue, parser.index)
+                                } else {
+                                    e
+                                }
+                            })
+                            .map(|n| match n {
                                 NumberAny::Int(NumberInt::Int(int)) => JsonValue::Int(int),
                                 #[cfg(feature = "num-bigint")]
                                 NumberAny::Int(NumberInt::BigInt(big_int)) => JsonValue::BigInt(big_int),
                                 NumberAny::Float(float) => JsonValue::Float(float),
-                            }
-                        }
+                            }),
                     };
 
-                    // now try to advance position in the current object
-                    if let Some(yet_another_key) = parser.object_step::<StringDecoder>(tape)?.map(create_cow) {
-                        // object continuing
-                        partial.insert(std::mem::replace(next_key, yet_another_key), value);
-                        peek = parser.peek()?;
-                        continue;
-                    }
+                    let object = match result {
+                        Ok(value) => {
+                            // now try to advance position in the current object
+                            match parser.object_step::<StringDecoder>(tape) {
+                                Ok(Some(yet_another_key)) => {
+                                    match parser.peek() {
+                                        Ok(next_peek) => {
+                                            // object continuing
+                                            partial.insert(
+                                                std::mem::replace(next_key, create_cow(yet_another_key)),
+                                                value,
+                                            );
+                                            peek = next_peek;
+                                            continue;
+                                        }
+                                        Err(e) if !(allow_partial && e.allowed_if_partial()) => return Err(e),
```

Comment:
Rather than having every possible `?` replaced in this way, I feel like there might have been a solution where the whole loop body goes into a sub-function, and then can match once on a possible error and if it's a partial end, wind up the recursion stack to build the partial datastructure.

But I'm a bit tired to think in more detail or push a useful commit to that effect tonight.